### ISSUE TITLE
Create new crate and set up basic infrastructure

### ISSUE THREAD

**Issue by marsella** [boltlabs-inc/key-mgmt#2]
Make a new crate (key-mgmt is not taken on crates.io, but we can always change the name later).

Set up binary - calling - library structure. Decide whether we want it to be like zeekoe, where the library modules were divided by functionality, and then by party, but the public module structure releases them by party::functionality.

Set up library with all the recommended warnings (copy from zeekoe, and maybe reference the Rust production book for any additional ones).

Set up CI to run all the Rust checks on PR to main (fmt, clippy, build, tests).

**Comment by gijsvl** [comment]
- Not sure if we already need to create a crate, I cannot really find how we've done that before. It seems someone becomes owner, maybe it's better to be @jakinyele .
- Not sure what you mean with "all the recommended warnings"
- I've setup 3 github workflows: 
-- default CI: runs on every PR, 
-- cargo audit (to check vulnerabilities in dependencies): runs once a day. 
-- integration tests: runs on every PR, but separately for performance reasons
- I've looked into tarpaulin, but I'm not sure if this is going to be very helpful at this point. There is an unstable feature in Rust that is going to do code coverage as well. Tarpaulin doesn't run on Mac for example, it is also not very clear where we are going to serve the results if we run it inside a github workflow. And what actions we are going to take based on the results. I think we just have to be vigilant for tests as part of every PR and that would give us better results at this point than code coverage.

**Comment by gijsvl** [comment]
I haven't created a PR at this point, given that everything in the repo is part of this issue anyway at this point. I also don't have the rights to change the repo settings in Github to protect the main branch etc.

**Comment by indomitableSwan** [comment]
> I haven't created a PR at this point, given that everything in the repo is part of this issue anyway at this point. I also don't have the rights to change the repo settings in Github to protect the main branch etc.

That will soon change, though. @marsella or @jakinyele do you have the rights to get the repo settings adjusted?

**Comment by indomitableSwan** [comment]
> * Not sure if we already need to create a crate, I cannot really find how we've done that before. It seems someone becomes owner, maybe it's better to be @jakinyele .

I think this is just referring to https://doc.rust-lang.org/book/ch07-00-managing-growing-projects-with-packages-crates-and-modules.html. i.e., we don't need to interact with crates.io at this point, but we want to manage the project with packages, crates, and modules. @marsella?

> * Not sure what you mean with "all the recommended warnings"
What you have looks OK to me.

> * I've setup 3 github workflows:
>   -- default CI: runs on every PR,
>   -- cargo audit (to check vulnerabilities in dependencies): runs once a day.
>   -- integration tests: runs on every PR, but separately for performance reasons
> * I've looked into tarpaulin, but I'm not sure if this is going to be very helpful at this point. There is an unstable feature in Rust that is going to do code coverage as well. Tarpaulin doesn't run on Mac for example, it is also not very clear where we are going to serve the results if we run it inside a github workflow. And what actions we are going to take based on the results. I think we just have to be vigilant for tests as part of every PR and that would give us better results at this point than code coverage.

Ah, yes, sounds like it won't be useful for us.

**Comment by marsella** [comment]
> Not sure if we already need to create a crate
Yes to what Swan said. We do not need to put it on crates.io yet, I think we would only do that when we have a "release" version.

> Not sure what you mean with "all the recommended warnings"

In libzkchannels-crypto, we annotated the library with specific things we wanted to get warnings on. I'm not sure how this interacts with the CI's invocation of clippy - are some of these "extras" that don't normally throw a warning?
https://github.com/boltlabs-inc/libzkchannels-crypto/blob/main/zkabacus-crypto/src/lib.rs#L22-L29

**Comment by jakinyele** [comment]
> I haven't created a PR at this point, given that everything in the repo is part of this issue anyway at this point. I also don't have the rights to change the repo settings in Github to protect the main branch etc.

I've updated the repo settings to protect the main branch.

**Comment by marsella** [comment]
A few other comments:
- Rename `integration-tests/` to just `tests/` and remove the pointer to it in the `Cargo.toml`. I think we should aim to use the default Rust conventions until the need arises, and avoid the non-standard-test-harness from zeekoe.
- Fix naming on e2e-tests. L28 says "Format check" but actually runs the tests

We may want to re-evaluate which tests run in the `e2e` flow and which run in `rust` once we have some tests and know which ones are slow. But I think the division is fine for now.

**Comment by gijsvl** [comment]
I've added the extra warnings. I think some of them might be a little overkill, but I guess it's good practice. I couldn't really find a good resource for what is recommended, but at least these are the same as Shea added to libzkchannels-crypto. I've removed unused_qualifications though, because it's more of a stylistic choice that I really don't care about.

@jakinyele I was just able to push to main, is the branch indeed protected?
